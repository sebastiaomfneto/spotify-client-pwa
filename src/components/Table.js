import React, { Children } from "react";

import "./Table.scss";

export default function Table({ columns = [], children }) {
  return (
    <div className="Table">
      <table>
        <thead>
          <tr>
            {columns.map((column, index) => (
              <th key={index}>{column}</th>
            ))}
          </tr>
        </thead>
        <tbody>
          {Children.count(children) > 0 ? (
            children
          ) : (
            <tr>
              <td colSpan={columns.length}>there are no records to display</td>
            </tr>
          )}
        </tbody>
      </table>
    </div>
  );
}
