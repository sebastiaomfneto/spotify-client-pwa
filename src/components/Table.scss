.Table {
  position: relative;
  overflow-y: auto;

  > table {
    width: 100%;
    border-collapse: collapse;
    white-space: nowrap;
    thead tr {
      line-height: 2.5rem;
      th {
        padding: 0 0.5rem;
        font-size: 0.75rem;
        text-transform: uppercase;
        color: #b3b3b3;
      }
    }

    tbody tr {
      border-bottom: 1px solid #666;
      line-height: 2.5rem;
      font-size: 0.85rem;
      &:hover {
        background-color: #333333;
      }

      td {
        padding: 0 0.5rem;
        text-align: center;
      }

      a {
        color: #fff;
        text-decoration: none;
        &:hover {
          text-decoration: underline;
        }
      }
    }
  }
}
