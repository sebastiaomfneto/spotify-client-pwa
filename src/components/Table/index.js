export { AlbumTable } from "./AlbumTable";
export { ArtistTable } from "./ArtistTable";
export { Table } from "./Table";
export { TrackTable } from "./TrackTable";
