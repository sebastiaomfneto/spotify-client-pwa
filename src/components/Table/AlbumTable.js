import React from "react";
import { Link } from "react-router-dom";

import { Table } from "./Table";
import { FavoriteButton } from "../FavoriteButton";

import { buildImageSrc } from "../../helpers";

function buildArtist(artists = []) {
  if (artists.length > 1) {
    return "Various artists";
  }

  return artists[0]?.name;
}

function buildAvailability(availableMarkets = []) {
  return availableMarkets.includes("BR")
    ? "Available in Brazil"
    : "Unavailable in Brazil";
}

export function AlbumTable({ albums = [] }) {
  return (
    <Table columns={["", "Image", "Name", "Artist", "Availability"]}>
      {albums.map((album) => (
        <tr key={album.id}>
          <td>
            <FavoriteButton id={album.id} />
          </td>
          <td>
            <img
              src={buildImageSrc(album.images)}
              alt="album"
              width="20px"
              height="20px"
            />
          </td>
          <td>
            <Link
              to={{
                pathname: `/albums/${album.id}`,
                state: { album },
              }}
            >
              {album.name}
            </Link>
          </td>
          <td>{buildArtist(album.artists)}</td>
          <td>{buildAvailability(album.available_markets)}</td>
        </tr>
      ))}
    </Table>
  );
}
