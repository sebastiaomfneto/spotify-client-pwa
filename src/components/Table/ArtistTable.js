import React from "react";
import { Link } from "react-router-dom";

import { Table } from "./Table";
import { FavoriteButton } from "../FavoriteButton";

import { buildImageSrc } from "../../helpers";

function buildGenresList(genres = []) {
  return genres.join(", ");
}

function buildPopularity(followers = 0) {
  switch (true) {
    case followers > 80:
      return "Hot";
    case followers >= 60 && followers <= 79:
      return "Cool";
    case followers >= 30 && followers <= 59:
      return "Regular";
    case followers < 30:
      return "Underground";
    default:
      return "";
  }
}

export function ArtistTable({ artists = [] }) {
  return (
    <Table columns={["", "Image", "Name", "Genres", "Popularity"]}>
      {artists.map((artist) => (
        <tr key={artist.id}>
          <td>
            <FavoriteButton id={artist.id} />
          </td>
          <td>
            <img
              src={buildImageSrc(artist.images)}
              alt="artist"
              width="20px"
              height="20px"
            />
          </td>
          <td>
            <Link
              to={{
                pathname: `/artists/${artist.id}`,
                state: { artist },
              }}
            >
              {artist.name}
            </Link>
          </td>
          <td>{buildGenresList(artist.genres)}</td>
          <td>{buildPopularity(artist.followers.total)}</td>
        </tr>
      ))}
    </Table>
  );
}
