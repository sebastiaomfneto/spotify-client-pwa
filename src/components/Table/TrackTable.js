import React from "react";

import { Table } from "./Table";
import { FavoriteButton } from "../FavoriteButton";

import { buildImageSrc } from "../../helpers";

function buildArtist(artists = []) {
  return artists.map((a) => a.name).join(", ");
}

function buildDuration(ms = 0) {
  const min = (ms / 1000 / 60) << 0;
  let sec = (ms / 1000) % 60 << 0;

  if (sec < 10) {
    sec = sec.toString() + "0";
  }

  return `${min}:${sec}`;
}

export function TrackTable({ album, tracks = [] }) {
  return (
    <Table columns={["", "Album Image", "Name", "Artist", "Album", "Duration"]}>
      {tracks.map((track) => (
        <tr key={track.id}>
          <td>
            <FavoriteButton id={track.id} />
          </td>
          <td>
            <img
              src={buildImageSrc(track.album?.images ?? album.images)}
              alt="track"
              width="20px"
              height="20px"
            />
          </td>
          <td>{track.name}</td>
          <td>{buildArtist(track.artists)}</td>
          <td>{track.album?.name ?? album.name}</td>
          <td>{buildDuration(track.duration_ms)}</td>
        </tr>
      ))}
    </Table>
  );
}
