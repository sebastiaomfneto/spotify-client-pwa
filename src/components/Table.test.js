import React from 'react';
import { render } from '@testing-library/react';

import { Table } from './Table';

describe('testing Table', () => {
  const columns = ['TEST_COLUMN_1'];
  const rows = ['TEST_ROW_1', 'TEST_ROW_2'];

  test('should match with snapshot', () => {
    const { container } = render(
      <Table columns={columns}>
        {rows.map((row) => (
          <tr key={row}>
            <td>{row}</td>
          </tr>
        ))}
      </Table>
    );

    expect(container).toMatchSnapshot();
  });
});
