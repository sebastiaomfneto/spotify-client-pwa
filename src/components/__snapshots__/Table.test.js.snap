// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing Table should match with snapshot 1`] = `
<div>
  <div
    class="Table"
  >
    <table>
      <thead>
        <tr>
          <th>
            TEST_COLUMN_1
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            TEST_ROW_1
          </td>
        </tr>
        <tr>
          <td>
            TEST_ROW_2
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
